### DISCUSSION TITLE
collect flux for the coarse mesh of a subapp from the results of the main app

### DISCUSSION THREAD

**Discussion by hityyds** [idaholab/moose#26144] [Q&A General] [answered]
I would like to realize CMFD (Coarse Mesh Finite Difference method) for transport calculations. It's an acceleration technique akin to multigrid. This method involves spatially weighted averaging of fine-grid iteration results from the main app to derive coarse-grid material properties. Is there a technique available for conducting flux statistics on cell faces and to obtain the flow on coarse-grid faces through area-weighting the flux values on fine-grid faces.

In the introduction to the transfer system, it is mentioned that spatial mapping between different grids can be accomplished using the transfer functionality. Consequently, is it possible to execute the entire statistical process of fine-grid iteration results and the weighted averaging process between coarse and fine grids within the transfer system? Or is it necessary to initially leverage AuxKernels or Postprocessors to conduct relevant statistics on the main app's results, followed by the spatial mapping (i.e., weighting and averaging of cells and cell faces) within the transfer system?

I acknowledge that my query may be intricate and extensive. Any guidance on any aspect of this matter would be highly appreciated.

**Comment by GiudGiud** [comment, marked as answer]
Hello

You ll have to make a user object derived from InternalSideUserObject or DomainUserObject to perform this accumulation in the internalSide routine

Guillaume

**Reply by hityyds** [reply]
Thank you for your guidance. Based on my understanding, InternalSideUserObject is designed for internal faces. Considering this, should I also create a SideUserObject derived class for handling boundary faces? Additionally, regarding the data transfer between coarse-mesh and fine-mesh, is there an existing transfer object in moose that facilitates the weighting and averaging process?

**Reply by GiudGiud** [reply]
Indeed. You could do that and have an internalSide object, a boundary side object etc

That s why we introduced the DomainUO that can do both sides, internal sides and elements

**Reply by GiudGiud** [reply]
None of the transfers in moose do the averaging and weighting for you. 

What we do is use UserObjects for that part then use the UOTransfer classes to get the user object values in the other application

**Reply by hityyds** [reply]
But how can we retain the averaged values for a coarse mesh? If we define a variable in the transport app to store these values, the DOFs of this variable will be significantly larger than necessary, as it is defined on the fine-mesh. For instance, with 1000 elements on the fine mesh and only 10 on the coarse mesh, the variables in the transport app might store DOFs for all 1000 elements, while we actually require only 10 DOFs for the 10 coarse elements. Am I misunderstanding this? Is it possible to define a variable based on the coarse mesh in the transport app, allowing the app to handle the fine mesh as its computational mesh?

**Reply by GiudGiud** [reply]
So yes, you would not use a variable on the fine mesh to hold any CMFD quantity. 

You can either use a multiapp and have a coarse mesh there

Or define your own arrays, like standard C++ vectors, in the user object. Not everything in a moose app has to be using moose objects, this is what user objects are for

**Reply by hityyds** [reply]
I understand, thank you very much.
